### PR TITLE
fix(mariner): use `advisory_id` for definition file names 

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
There are times when CBL-Mariner uses multiple `definitions` for a single CVE.
e.g:
```
<definition class="vulnerability" id="oval:com.microsoft.cbl-mariner:def:31880" version="1">
      <metadata>
        <title>CVE-2023-5678 affecting package openssl for versions less than 1.1.1k-28</title>
        <affected family="unix">
          <platform>CBL-Mariner</platform>
        </affected>
        <reference ref_id="CVE-2023-5678" ref_url="https://nvd.nist.gov/vuln/detail/CVE-2023-5678" source="CVE"/>
        <patchable>true</patchable>
        <advisory_id>31880-1</advisory_id>
        <severity>Medium</severity>
        <description>CVE-2023-5678 affecting package openssl for versions less than 1.1.1k-28. A patched version of the package is available.</description>
      </metadata>
      <criteria operator="AND">
        <criterion comment="Package openssl is earlier than 1.1.1k-28, affected by CVE-2023-5678" test_ref="oval:com.microsoft.cbl-mariner:tst:31880000"/>
      </criteria>
    </definition>
    <definition class="vulnerability" id="oval:com.microsoft.cbl-mariner:def:31872" version="1">
      <metadata>
        <title>CVE-2023-5678 affecting package edk2 for versions less than 20230301gitf80f052277c8-38</title>
        <affected family="unix">
          <platform>CBL-Mariner</platform>
        </affected>
        <reference ref_id="CVE-2023-5678" ref_url="https://nvd.nist.gov/vuln/detail/CVE-2023-5678" source="CVE"/>
        <patchable>true</patchable>
        <advisory_id>31872-1</advisory_id>
        <severity>Medium</severity>
        <description>CVE-2023-5678 affecting package edk2 for versions less than 20230301gitf80f052277c8-38. A patched version of the package is available.</description>
      </metadata>
      <criteria operator="AND">
        <criterion comment="Package edk2 is earlier than 20230301gitf80f052277c8-38, affected by CVE-2023-5678" test_ref="oval:com.microsoft.cbl-mariner:tst:31872000"/>
      </criteria>
    </definition>
```
To avoid overwriting - use `advisory_id` field for file names.
See https://github.com/aquasecurity/vuln-list-update/pull/271#issuecomment-2111678641

## Related Issues
- https://github.com/aquasecurity/trivy-db/issues/379

### CODE REVIEW

**Comment by knqyf263** [comment]
I'm checking with the Aqua commercial team.

**Comment by eric-desrochers** [comment]
Any ETA for this PR to be merged in Trivy ?

**Comment by eric-desrochers** [comment]
@knqyf263 any development based on your check with Aqua commercial team ?

**Comment by knqyf263** [comment]
I'll remind them

**Comment by eric-desrochers** [comment]
> I'll remind them

@knqyf263 any update ? This is impacting AKS Image Cleaner in its ability to retire unsecure images for which CVE are found.

**Comment by knqyf263** [comment]
I reminded them again, but they seem to be busy. If you're in a hurry, you can update your data as described below.
https://github.com/aquasecurity/trivy-db/issues/379#issuecomment-1941447343

**Comment by knqyf263** [comment]
@DmitriyLewen Can't we save two files named the advisory id instead of CVE-ID, like `/definitions/2023/31872-1.json`?

**Comment by DmitriyLewen** [comment]
hm... i didn't think about this.

I see 2 points:
- I checked oval files. Looks like all `definitions` contain `advisory_id`. Also all `advisory_id` are unique.
  ```bash
  ➜ cat cbl-mariner-2.0-oval.xml| grep advisory_id | sort | uniq -d   
  ➜ cat cbl-mariner-1.0-oval.xml| grep advisory_id | sort | uniq -d
  ➜ cat cbl-mariner-1.0-oval.xml| grep '<definition class="vulnerability"' | wc -l
      2252
  ➜ cat cbl-mariner-1.0-oval.xml| grep advisory_id | wc -l         
      2252
  ➜ cat cbl-mariner-2.0-oval.xml| grep '<definition class="vulnerability"' | wc -l
      2702
  ➜ cat cbl-mariner-2.0-oval.xml| grep advisory_id | wc -l
      2702
  ```
  But `advisory_id` is custom field - https://github.com/OVALProject/Language/blob/7fa7bba7b48f09decb732d00b2be032a487ff9fc/schemas/oval-definitions-schema.xsd#L242-L276. Therefore, we cannot be 100% sure about this field.
- Working with these files will be more difficult. `advisory_id` means nothing to user. Therefore, to find file for CVE, you need to use search. It's not a big problem, but it's still 1 extra action.

**Comment by knqyf263** [comment]
>But advisory_id is custom field - https://github.com/OVALProject/Language/blob/7fa7bba7b48f09decb732d00b2be032a487ff9fc/schemas/oval-definitions-schema.xsd#L242-L276. Therefore, we cannot be 100% sure about this field.

`id` and `version` are required. I think we can construct an advisory id from them.
https://github.com/OVALProject/Language/blob/7fa7bba7b48f09decb732d00b2be032a487ff9fc/schemas/oval-definitions-schema.xsd#L237-L238

```
    <definition class="vulnerability" id="oval:com.microsoft.cbl-mariner:def:31880" version="1">
```

>Working with these files will be more difficult. advisory_id means nothing to user. Therefore, to find file for CVE, you need to use search. It's not a big problem, but it's still 1 extra action.

It's no big deal. We already use custom advisory IDs for others.
https://github.com/aquasecurity/vuln-list/tree/main/ghsa/go/github.com/aws/aws-sdk-go
https://github.com/aquasecurity/vuln-list/tree/main/alma/9/2024

**Comment by DmitriyLewen** [comment]
hm... looks like your are right.
IIUC in some cases `mariner` uses only latest from `version`:
```
    <definition class="vulnerability" id="oval:com.microsoft.cbl-mariner:def:27423" version="2000000001">
...
        <advisory_id>27423-1</advisory_id>
```

But it doesn't have much impact.

**Comment by knqyf263** [comment]
>IIUC in some cases mariner uses only latest from version:

What do you mean? `2000000001` is the latest?

**Comment by DmitriyLewen** [comment]
hm... i missed word 😞 
I meant `mariner uses only latest number from version`
`version="2000000001"` -> 1
`version="2000000002"` -> 2
`version="2000000000" -> without version

**Comment by knqyf263** [comment]
Got it, you mean they use the last digit. Even `27423-2000000001` is not bad since we just need a unique identifier.

**Comment by DmitriyLewen** [comment]
only cbl-1.0 uses version `2000..`.
cbl-2.0 uses single digit as version (1,2,3, etc.).
I think we can reproduce their logic.

**Comment by knqyf263** [comment]
If we go with this approach, what changes do we need in trivy-db compared to [the current PR](https://github.com/aquasecurity/trivy-db/pull/382)?

**Comment by DmitriyLewen** [comment]
we will only need to update names of test files to match the names in `vuln-list`.

File: mariner/mariner.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ import (
 )
 
 const (
-	repoURL = "https://github.com/microsoft/CBL-MarinerVulnerabilityData/archive/refs/heads/main.tar.gz//CBL-MarinerVulnerabilityData-main"
+	repoURL = "https://github.com/microsoft/AzureLinuxVulnerabilityData/archive/refs/heads/main.tar.gz//AzureLinuxVulnerabilityData-main"
```
Comment (DmitriyLewen) [inline mariner/mariner.go:20]: IIUC Microsoft renamed repository:
`CBL-MarinerVulnerabilityData` -> `AzureLinuxVulnerabilityData`

File: mariner/mariner.go [hunk 2]
```diff
@@ -154,8 +152,24 @@ func (c Config) update(version, path string) error {
...
+		if def.Version != "" && def.Version[len(def.Version)-1:] != "0" {
+			advisoryID = fmt.Sprintf("%s_%s", advisoryID, def.Version[len(def.Version)-1:])
+		}
+	}
+	// Use advisory_id for file name to avoid overwriting files when there are 2 definitions for same CVE
+	// cf. https://github.com/aquasecurity/trivy-db/issues/379
+	fileName := fmt.Sprintf("%s.json", advisoryID)
 
-func (c Config) saveAdvisoryPerYear(dirName string, vulnID string, def Definition) error {
+	vulnID := def.Metadata.Reference.RefID
```
Comment (DmitriyLewen) [inline mariner/mariner.go:172]: I thought about using `advisory_date` field.
But this field does not always exist:
```
➜ cat cbl-mariner-1.0-oval.xml| grep ' <definition class="vulnerability"' | toilet -l
      2252
➜ cat cbl-mariner-1.0-oval.xml| grep '<advisory_date>' | toilet -l
      2070
```
So I'm leaving logic with year number from CVE.

**Comment by DmitriyLewen** [comment]
@knqyf263 I updated this PR. Can you take a look?

**Comment by knqyf263** [comment]
Even after we merge this PR, will `trivy-db` not fail?

**Comment by DmitriyLewen** [comment]
right.
I used mariner dir with new file names:
```
➜  make db-build
./trivy-db build --cache-dir cache --update-interval 6h
2024/05/15 15:51:45 Updating vulnerability database...
2024/05/15 15:51:45 Updating cbl-mariner data...
2024/05/15 15:51:45     Parsing cache/vuln-list/mariner/1.0
2024/05/15 15:51:45     Parsing cache/vuln-list/mariner/2.0
```

**Comment by knqyf263** [comment]
Thanks for confirming. I'll merge this PR, then.

File: mariner/mariner.go [hunk 3]
```diff
@@ -154,8 +152,24 @@ func (c Config) update(version, path string) error {
 
 	return nil
 }
+func (c Config) saveAdvisoryPerYear(dirName string, def Definition) error {
+	// Mariner uses `<ID>_<last_number_from_version>` format for `advisory_id`.
```
Comment (knqyf263) [inline mariner/mariner.go:156]: AFAIK, they use hyphens.

```suggestion
	// Mariner uses `<ID>-<last_number_from_version>` format for `advisory_id`.
```

Comment (knqyf263) [reply mariner/mariner.go:156]: We should start calling Azure Linux.

```suggestion
	// Azure Linux uses `<ID>_<last_number_from_version>` format for `advisory_id`.
```

File: mariner/mariner.go [hunk 4]
```diff
@@ -154,8 +152,24 @@ func (c Config) update(version, path string) error {
...
+	// But `advisory_id` is not required field.
+	// Therefore, if `advisory_id` is not exist, we create this field independently.
+	// cf. https://github.com/aquasecurity/vuln-list-update/pull/271#issuecomment-2111678641
+	advisoryID := def.Metadata.AdvisoryID
+	if advisoryID == "" {
+		advisoryID = def.ID
```
Comment (knqyf263) [inline mariner/mariner.go:162]: The ID is something like `oval:com.microsoft.cbl-mariner:obj:31880001`. Shouldn't we extract the last digits, `31880001`?

File: mariner/mariner.go [hunk 5]
```diff
@@ -154,8 +152,24 @@ func (c Config) update(version, path string) error {
...
+		// for `0` versions `_0` suffix is omitted.
+		if def.Version != "" && def.Version[len(def.Version)-1:] != "0" {
+			advisoryID = fmt.Sprintf("%s_%s", advisoryID, def.Version[len(def.Version)-1:])
```
Comment (knqyf263) [inline mariner/mariner.go:165]:
```suggestion
			advisoryID = fmt.Sprintf("%s-%s", advisoryID, def.Version[len(def.Version)-1:])
```

File: mariner/mariner.go [hunk 6]
```diff
@@ -154,8 +152,24 @@ func (c Config) update(version, path string) error {
 
 	return nil
 }
+func (c Config) saveAdvisoryPerYear(dirName string, def Definition) error {
+	// Mariner uses `<ID>_<last_number_from_version>` format for `advisory_id`.
+	// But `advisory_id` is not required field.
+	// Therefore, if `advisory_id` is not exist, we create this field independently.
```
Comment (knqyf263) [inline mariner/mariner.go:158]: nit

```suggestion
	// Therefore, if `advisory_id` does not exist, we create this field independently.
```

**Review by knqyf263 — COMMENTED** [review]
I left some small comments.

File: mariner/mariner.go (hunk 3 above)
Comment (DmitriyLewen) [reply mariner/mariner.go]: fixed in https://github.com/aquasecurity/vuln-list-update/pull/271/commits/995ea97f301c04bacd37987086c6d8992add1703

File: mariner/mariner.go (hunk 4 above)
Comment (DmitriyLewen) [reply mariner/mariner.go]: Nice catch!
Changed in https://github.com/aquasecurity/vuln-list-update/pull/271/commits/995ea97f301c04bacd37987086c6d8992add1703
Also added tests - https://github.com/aquasecurity/vuln-list-update/pull/271/commits/7074604717e7a6a93d6a4a689d4098cc9b7268bd

File: mariner/mariner.go (hunk 5 above)
Comment (DmitriyLewen) [reply mariner/mariner.go]: fixed in https://github.com/aquasecurity/vuln-list-update/pull/271/commits/995ea97f301c04bacd37987086c6d8992add1703

File: mariner/mariner.go (hunk 6 above)
Comment (DmitriyLewen) [reply mariner/mariner.go]: updated in https://github.com/aquasecurity/vuln-list-update/pull/271/commits/995ea97f301c04bacd37987086c6d8992add1703

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
Triggered the workflow
https://github.com/aquasecurity/vuln-list-update/actions/runs/9095013101

**Pull request merged**